File: src/main/kotlin/icu/takeneko/libNeko/nyatwork/Reader.java
package icu.takeneko.libNeko.nyatwork;

import java.util.Optional;
import java.util.function.Function;

@FunctionalInterface
public interface Reader<T> extends Function<FriendlyByteBuf, T> {
    default Reader<Optional<T>> asOptional() {
        return (buf) -> {
            return buf.readOptional(this);
        };
    }
}

File: src/main/kotlin/icu/takeneko/libNeko/nyatwork/util/Reader.kt
package icu.takeneko.libNeko.nyatwork.util

import java.util.*
import java.util.function.Function

fun interface Reader<T: Any> : Function<FriendlyByteBuf, T> {
    fun asOptional(): Reader<Optional<T>> {
        return Reader { buf: FriendlyByteBuf ->
            buf.readOptional(
                this
            )
        }
    }
}